src/cargo/ops/cargo_package.rs

```diff
@@ -42,6 +44,43 @@ pub fn package(ws: &Workspace, opts: &PackageOpts) -> CargoResult<Option<FileLoc
 
     verify_dependencies(pkg)?;
 
+    // `list_files` outputs warnings as a side effect, so only do it once.
+    let src_files = src.list_files(pkg)?;
+
+    // Check (git) repository state, getting the current commit hash if not
+    // dirty. This will `bail!` if dirty, unless allow_dirty.
+    let sha1_hash = check_repo_state(pkg, &src_files, opts.allow_dirty)?;
+
+    // Write out VCS info file for package, if we have a sha1 hash, then
+    // preserve a shared lock on the file.
+    let vcs_info = if let Some(hsh) = sha1_hash {
+        let mut ifile = ws.target_dir().open_rw(
+            VCS_INFO_FILE,
+            config,
+            "(git) VCS info"
+        )?;
+        ifile.file().set_len(0)?;
+        // Manual JSON format is safe given sha1 hash is hex encoded.
+        writeln!(ifile, r#"{{"git_head_revision_sha1":"{}"}}"#, hsh)?;
+        ifile.flush()?;
+        ifile.file().sync_all()?;
```

Comment:
Hm could you elaborate a bit on why the file locks and syncs are needed here? I was thinking we'd just synthesize this file directly into the tarball (like `Cargo.toml.orig` without actually putting it on the filesystem)

Reply:
I figured early on that I wanted a sentry file to try and make it at least *inconvenient* to spoof the hash, for example by protecting against interference from another `cargo package` process.  The changes would be reduced without the sentry file however. Do you the anti-spoof efforts are unwarranted?